### PR TITLE
Closes #571 | Add/Update Dataloader UP2.0

### LINKED ISSUE #571: Create dataset loader for UP2.0
Dataloader name: `up2/up2.py`
DataCatalogue: http://seacrowd.github.io/seacrowd-catalogue/card.html?up2

|     Dataset| up2 |
|-------------|---|
| Description | Southeast Asian language subsets from Universal Propositions (UP) 2.0 dataset. Semantic role labeling (SRL) is a shallow semantic parsing task that identifies “who did what to whom when, where etc” for each predicate in a sentence. It provides an intermediate (shallow) level of a semantic representation that helps the map from syntactic parse structures to more fully-specified representations of meaning. |
| Subsets | ind, vie |
| Languages | ind, vie |
| Tasks | Semantic Role Labeling |
| License | Community Data License Agreement – Permissive, Version 1.0 (cdla-permissive-1.0) |
| Homepage | https://github.com/UniversalPropositions/UP_Indonesian-GSD, https://github.com/UniversalPropositions/UP_Vietnamese-VTB |
| HF URL | - |
| Paper URL | https://aclanthology.org/2022.lrec-1.181.pdf |

### PR DESCRIPTION
**Pull request by fhudi**
Closes #571 

### Checkbox
- [x] Confirm that this PR is linked to the dataset issue.
- [x] Create the dataloader script `seacrowd/sea_datasets/{my_dataset}/{my_dataset}.py` (please use only lowercase and underscore for dataset folder naming, as mentioned in dataset issue) and its `__init__.py` within `{my_dataset}` folder.
- [x] Provide values for the `_CITATION`, `_DATASETNAME`, `_DESCRIPTION`, `_HOMEPAGE`, `_LICENSE`, `_LOCAL`, `_URLs`, `_SUPPORTED_TASKS`, `_SOURCE_VERSION`, and `_SEACROWD_VERSION` variables.
- [x] Implement `_info()`, `_split_generators()` and `_generate_examples()` in dataloader script.
- [ ] Make sure that the `BUILDER_CONFIGS` class attribute is a list with at least one `SEACrowdConfig` for the source schema and one for a seacrowd schema.
- [x] Confirm dataloader script works with `datasets.load_dataset` function.
- [x] Confirm that your dataloader script passes the test suite run with `python -m tests.test_seacrowd seacrowd/sea_datasets/<my_dataset>/<my_dataset>.py` or `python -m tests.test_seacrowd seacrowd/sea_datasets/<my_dataset>/<my_dataset>.py --subset_id {subset_name_without_source_or_seacrowd_suffix}`.
- [ ] If my dataset is local, I have provided an output of the unit-tests in the PR (please copy paste). This is OPTIONAL for public datasets, as we can test these without access to the data files.

---

Currently implemented as source-only, as there seems to be no matching task.
(Based on the information provided in the card)

Issue was initially raised in the [issue thread #571](https://github.com/SEACrowd/seacrowd-datahub/issues/571#issuecomment-2073953395).

### CODE REVIEW

**Review by SamuelCahyawijaya — APPROVED** [review]
@fhudi : LGTM!

File: seacrowd/sea_datasets/up2/up2.py
Comment (ljvmiranda921) [inline seacrowd/sea_datasets/up2/up2.py:84]:
```suggestion
```

File: seacrowd/sea_datasets/up2/up2.py
Comment (ljvmiranda921) [inline seacrowd/sea_datasets/up2/up2.py:203]:
```suggestion
```

File: seacrowd/sea_datasets/up2/up2.py
Comment (ljvmiranda921) [inline seacrowd/sea_datasets/up2/up2.py:116]: I assume this one doesn't have a SEACrowd schema. I guess we can delete this commented block?

File: seacrowd/sea_datasets/up2/up2.py
Comment (ljvmiranda921) [inline seacrowd/sea_datasets/up2/up2.py:139]:
```suggestion
```

**Review by ljvmiranda921 — APPROVED** [review]
Ran the data loader and worked! I guess this one doesn't have a SEACrowd schema, no? Some commented blocks that can be deleted (optionally). But good to merge as is 👍

**Comment by khelli07** [comment]
Changed ` utils/common_parser.py > load_ud_data`
```
dataset_raw = parse(open(filepath).read())
```
to
```
with open(filepath, encoding='utf-8', errors='replace') as file:
    dataset_raw = parse(file.read())
```

But please confirm if these result are correct or not
1. Ran through loader `if __name__ == "__main__"`:
image

2. A test sample:
image

**Comment by holylovenia** [comment]
Hi @SamuelCahyawijaya @ljvmiranda921 @khelli07 and @fhudi, I won't merge this dataloader as-is due to the error reported by @khelli07. Let's wait until @fhudi addresses @ljvmiranda921 and @khelli07's feedback.

**Comment by fhudi** [comment]
> Unfortunately the test did not pass on my end. Command I ran: `python -m tests.test_seacrowd seacrowd/sea_datasets/up2/up2.py --subset_id up2`

@khelli07 
Note that this is source-only dataloader.

As mentioned in the [reviewing checklist point 3](https://github.com/SEACrowd/seacrowd-datahub/blob/master/REVIEWING.md#dataloader-check), you need to execute test script using the one that is appropriate under the [test folder](https://github.com/SEACrowd/seacrowd-datahub/tree/master/tests).
Please use the one for [source_only](https://github.com/SEACrowd/seacrowd-datahub/blob/master/tests/test_seacrowd_source_only.py).

You can rever to [the example of executable test script](https://github.com/SEACrowd/seacrowd-datahub/blob/master/test_example_source_only.sh) on how to use it.

---

P.S.:
@holylovenia 
I can't find any readme file referring to how to test source-only dataset. Maybe lost when migrating from NusaCrowd 🥲.

File: seacrowd/sea_datasets/up2/up2.py
Comment (fhudi) [reply seacrowd/sea_datasets/up2/up2.py]: done. 👍

**Comment by fhudi** [comment]
> Changed ` utils/common_parser.py > load_ud_data`
> 
> ```
> dataset_raw = parse(open(filepath).read())
> ```
> 
> to
> 
> ```
> with open(filepath, encoding='utf-8', errors='replace') as file:
>     dataset_raw = parse(file.read())
> ```
> 
> But please confirm if these result are correct or not

@khelli07 
Thanks for the suggestion. LGTM.
But may I know the concise reasoning on these changes?

**Comment by khelli07** [comment]
Hi, the changes is to avoid this error:
`UnicodeDecodeError: 'charmap' codec can't decode byte 0x9d in position 282814: character maps to <undefined>`

It appears when I ran `datasets.load_dataset` too.

With that change, I successfully ran `datasets.load_dataset` and `python -m tests.test_seacrowd_source_only seacrowd/sea_datasets/up2/up2.py --subset_id up2`.

**Comment by khelli07** [comment]
Nope, I use Windows without WSL. Let me try it.

**Comment by khelli07** [comment]
Ok just finished testing, both of them works. So you might be able to choose either one.

**Comment by holylovenia** [comment]
Hi @fhudi @khelli07, I would like to let you know that we plan to finalize the calculation of the open contributions (e.g., dataloader implementations) **in 31 hours**, so it'd be great if we could wrap up the reviewing and merge this PR before then.

Has everything been addressed now? Can we merge this PR?

**Comment by ljvmiranda921** [comment]
I think this is just waiting for @khelli07 's review. But from the last comment it seems that it has been resolved 🤔 . I was able to run this so I think it's mostly a platform-specific issue. Dataloader works though so I think it's safe to merge!

**Review by khelli07 — CHANGES_REQUESTED** [review]
The `load_ud_data` in the `common_parser.py` hasn't been changed yet. Without that change, I don't think I can run successfully in my Windows. 

And yes, as @ljvmiranda921 said, I agree that this is a platform-specific issue. If that can be tolerated, then can merge.

**Comment by fhudi** [comment]
@khelli07 
Sorry for the late reply.
Honestly, adding the parameter `errors='replace'` might change the overall behaviour.
As this function is used by other DataLoader as well, I am a bit skeptical in including this change.
Also, based on your reply, the root cause does not seem to be from the additional parameters.
However, I have made the adjustment as seen from [the commit](https://github.com/SEACrowd/seacrowd-datahub/pull/660/commits/91d389c8d3a78982de6fb76739d028ffdce78b3d).
cc: @holylovenia @ljvmiranda921

**Review by khelli07 — APPROVED** [review]
LGTM!

**Comment by fhudi** [comment]
@khelli07 @ljvmiranda921
I believe this is save for merging?

**Comment by sabilmakbar** [comment]
> open(filepath, encoding='utf-8')
> open(filepath, errors='replace')

Sorry for jumping it late. I think this is a platform-based issue where, for non-Linux/MacOS, the encoding doesn't default to `utf-8`. Hence, the first one is preferable.

**Comment by sabilmakbar** [comment]
Since both reviewers have given the approver (and the dataloader assignee best not to merge it themselves even though it has been approved), I'll proceed with the merging.

fyi @khelli07 @ljvmiranda921

**Pull request merged**
